tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/LabelPropagationTest.java

```diff
@@ -0,0 +1,94 @@
+package org.palladiosimulator.dataflow.confidentiality.analysis;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.palladiosimulator.dataflow.confidentiality.analysis.AnalysisUtils.assertCharacteristicAbsent;
+import static org.palladiosimulator.dataflow.confidentiality.analysis.AnalysisUtils.assertCharacteristicPresent;
+
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+public class LabelPropagationTest extends AnalysisFeatureTest {
+
+    /**
+     * Tests the present characteristics of the found action sequences
+     * <p>
+     * Fails if the analysis does not propagate the correct characteristics to each ActionSequence
+     */
+    @ParameterizedTest(name = "{index}. {3}.{4}.{5} present at index {2} in action sequence")
+    @DisplayName("Sequence elements should have correct characteristics present")
+    @MethodSource("characteristicsPresentProvider")
+    public void characteristicsPresentTest(StandalonePCMDataFlowConfidentialtyAnalysis analysis, int sequenceIndex,
+            int elementIndex, String variable, String characteristicType, String characteristicValue) {
+        var sequences = analysis.findAllSequences();
+        var propagationResult = analysis.evaluateDataFlows(sequences);
+
+        assertTrue(propagationResult.size() >= sequenceIndex);
+
+        assertCharacteristicPresent(propagationResult.get(sequenceIndex), elementIndex, variable, characteristicType,
+                characteristicValue);
+    }
+
+    /**
+     * Provides test data to the {@code characteristicsPresentTest} Test class.
+     * <p>
+     * Each invocation contains the index into the list of sequences and a map which maps a index
+     * into the action sequence to the expected characteristic
+     * 
+     * @return Returns a stream of test data used for each invocation
+     */
+    private Stream<Arguments> characteristicsPresentProvider() {
+        return Stream.of(Arguments.of(travelPlannerAnalysis, 0, 0, "ccd", "GrantedRoles", "User"),
```

Comment:
Same comment about readability like in the other PR :D

Reply:
Also: It would be great to have a good toString() representation or console log of the propagated labels that prints all the information that is tested here. I will add an issue for it